### PR TITLE
Add support for the demux endpoint

### PR DESCRIPTION
**Pull request by katherine-epp**
Adds support for the `demux` endpoint added in Clarity 5.1 and a helper method that uses it to find the inputs to a pool or pool of pools.

The `get_pool_inputs` function and the objects and properties it uses are used in the sample sheet code I'm currently working on and were mainly tested by running that script. The rest of the properties and functions in the `ArtifactDemux` and `DemuxArtifact` classes were added for completeness and have been only tested by the automated tests I added. 

Most of my testing has been on Python 2.7, but I did run the automated tests and also my sample sheet script with 3.8 to make sure there weren't any obvious issues.

### CODE REVIEW

**Review by mgillis — APPROVED** [review]
I'm not sure about the class names ArtifactDemux and DemuxArtifact, but I don't have any better ideas. I didn't look at the spec to ensure this is correct, but code-wise this looks good to me. I would recommend someone else also take a look who has looked at this all more recently than me too.

**Comment by katherine-epp** [comment]
Thanks, Mike.

> I'm not sure about the class names ArtifactDemux and DemuxArtifact, but I don't have any better ideas.

`ArtifactDemux` seemed analogous to `StepActions`,`StepDetails`, `StepPools`, etc., so I didn't think about alternatives for that one. I was using `PoolArtifact` instead of `DemuxArtifact` at one point, but switched to bring the naming in line with the Clarity API documentation. I can change it back if `PoolArtifact` seems like a better name to you.

**Comment by Mark-Swinkels** [comment]
In general we have tried to keep our naming quite close to the Clarity
naming to minimize our need to document the details of things.

-- Mark

[image: Logo] Mark Swinkels CTO
mark.swinkels@semaphoresolutions.ca
1 250 896 6538
1 844 744 3577
semaphoresolutions.com <http://www.semaphoresolutions.com>
This message and any attachments are confidential and intended for the
named addressee(s) only. If you have received this message in error, please
notify immediately the sender, then delete the message. Any unauthorized
modification, edition, use or dissemination is prohibited. The sender shall
not be liable for this message if it has been modified, altered, falsified,
infected by a virus or even edited or disseminated without authorization.

**Comment by katherine-epp** [comment]
While I was doing my final round of testing on the script I implemented this for, I discovered that all of the test data I used while developing this feature hit an edge case where the demux endpoint returns less information than it does in more standard cases.

The pools I used for testing contained only artifacts from the same sample (yeah... I really should have known better ;)), so what I saw was something like:

<details><summary>pool of pools that contain multiple artifacts from one sample</summary><p>

```xml
<art:demux xmlns:art="http://genologics.com/ri/artifact" uri="https://qalocal/api/v2/artifacts/2-1024/demux">
    <artifact uri="https://qalocal/api/v2/artifacts/2-1024" name="1S2A-3S3A"/>
    <demux>
        <pool-step uri="https://qalocal/api/v2/steps/122-553" name="Combine Pools Step"/>
        <artifacts>
            <artifact uri="https://qalocal/api/v2/artifacts/2-1017" name="1S2A">
                <samples>
                    <sample uri="https://qalocal/api/v2/samples/EPP1A6" limsid="EPP1A6"/>
                </samples>
                <reagent-labels>
                    <reagent-label name="Reagent Label 1 (ACGT-TGCA)"/>
                    <reagent-label name="Reagent Label 2 (CGTA-ATGC)"/>
                </reagent-labels>
            </artifact>
            <artifact uri="https://qalocal/api/v2/artifacts/2-1018" name="1S2A">
                <!-- similar output to above -->
            </artifact>
        </artifacts>
    </demux>
</art:demux>
```
</p></details>

When I tested my script, I happened to pool artifacts from different samples together and discovered that demux artifacts can have demux details blocks as well as samples and reagent labels:

<details><summary>pool of pools that contain artifacts from multiple samples</summary><p>

```xml
<art:demux xmlns:art="http://genologics.com/ri/artifact" uri="https://qalocal/api/v2/artifacts/2-1024/demux">
    <artifact uri="https://qalocal/api/v2/artifacts/2-1024" name="1S2A-3S3A"/>
    <demux>
        <pool-step uri="https://qalocal/api/v2/steps/122-553" name="Combine Pools Step"/>
        <artifacts>
            <artifact uri="https://qalocal/api/v2/artifacts/2-1019" name="3S3A">
                <samples>
                    <sample uri="https://qalocal/api/v2/samples/EPP1A13" limsid="EPP1A13"/>
                    <sample uri="https://qalocal/api/v2/samples/EPP1A15" limsid="EPP1A15"/>
                    <sample uri="https://qalocal/api/v2/samples/EPP1A16" limsid="EPP1A16"/>
                </samples>
                <reagent-labels>
                    <reagent-label name="Reagent Label 1 (ACGT-TGCA)"/>
                    <reagent-label name="Reagent Label 2 (CGTA-ATGC)"/>
                    <reagent-label name="Reagent Label 3 (GTAC-CATG)"/>
                </reagent-labels>
                <demux>
                    <pool-step uri="https://qalocal/api/v2/steps/122-552" name="Pooling Step"/>
                    <artifacts>
                        <artifact uri="https://qalocal/api/v2/artifacts/2-632" name="20201119-2-5">
                            <samples>
                                <sample uri="https://qalocal/api/v2/samples/EPP1A13" limsid="EPP1A13"/>
                            </samples>
                            <reagent-labels>
                                <reagent-label name="Reagent Label 1 (ACGT-TGCA)"/>
                            </reagent-labels>
                        </artifact>
                        <artifact uri="https://qalocal/api/v2/artifacts/2-636" name="20201119-2-7">
                            <samples>
                                <sample uri="https://qalocal/api/v2/samples/EPP1A15" limsid="EPP1A15"/>
                            </samples>
                            <reagent-labels>
                                <reagent-label name="Reagent Label 2 (CGTA-ATGC)"/>
                            </reagent-labels>
                        </artifact>
                        <artifact uri="https://qalocal/api/v2/artifacts/2-638" name="20201119-2-8">
                            <samples>
                                <sample uri="https://qalocal/api/v2/samples/EPP1A16" limsid="EPP1A16"/>
                            </samples>
                            <reagent-labels>
                                <reagent-label name="Reagent Label 3 (GTAC-CATG)"/>
                            </reagent-labels>
                        </artifact>
                    </artifacts>
                </demux>
            </artifact>
            <artifact uri="https://qalocal/api/v2/artifacts/2-1018" name="3S3A">
                <!-- similar output to above -->
            </artifact>
        </artifacts>
    </demux>
</art:demux>
```
</p></details>

The result of this is that my implementation of the data model is incomplete since it doesn't parse the nested `demux` element and that my helper function (`get_pool_inputs`) makes far more API calls than it needs to in most cases (generally, it should only need to get the demux for the pool, but currently it also gets the demuxes for all of the pools and artifacts it contains).

Obviously, I will need to complete the data model before this can be merged, but I'm not quite sure what to do about `get_pool_inputs`. For my own project, I will need to stick with the current approach unless I can confirm that the script will never be used with a pool containing artifacts from only one sample. That doesn't seem appropriate for the library, though, so I'm inclined to move the helper back to my script and just leave the data model in the library.

File: s4/clarity/artifact.py [hunk 1]
```diff
@@ -213,5 +214,89 @@ def reagent_label_name(self, reagent_label_name):
...
+
+
+class DemuxArtifact(WrappedXml):
+    """
+    Corresponds to the 'demux-artifact' type in the 'artifact' namespace of the Clarity API.
+    """
+    artifact = subnode_link(Artifact, ".")
+    reagent_labels = subnode_element_list(ReagentLabel, "reagent-labels", "reagent-label", readonly=True)
+
+    @property
```
[the remark this answers is not in the document]
Comment (Mark-Swinkels) [reply s4/clarity/artifact.py:241]: Don't think it matters

File: s4/clarity/artifact.py [hunk 2]
```diff
@@ -213,5 +214,89 @@ def reagent_label_name(self, reagent_label_name):
...
+    def demux(self):
+        # type: () -> DemuxDetails
+        """
+        The element `DemuxDetails` from subnode 'demux'
+
+        :type: DemuxDetails
+        """
+        demux_root = self.xml_root.find("./demux")
+        if (demux_root is None):
+            if (len(self.samples) == 1 and len(self.reagent_labels) > 1):
```
Comment (Mark-Swinkels) [inline s4/clarity/artifact.py:251]: Should probably be a comment why this is in here. In the moment we both know but I suspect it will be confusing to future us or anyone else.

**Review by Mark-Swinkels — APPROVED** [review]
Looks good

File: s4/clarity/artifact.py (hunk 2 above)
Comment (katherine-epp) [reply s4/clarity/artifact.py:251]: Good point. I'll add one.

Comment (katherine-epp) [reply s4/clarity/artifact.py:256]: While I was writing this, I realized that our workaround doesn't cover the case where the sub-pool contains one labeled artifact and one unlabeled artifact from the same sample. Might wait for someone to report that that's actually a problem, though. :)

File: s4/clarity/artifact.py [hunk 3]
```diff
@@ -213,5 +214,100 @@ def reagent_label_name(self, reagent_label_name):
...
+    def demux(self):
+        # type: () -> DemuxDetails
+        """
+        The element `DemuxDetails` from subnode 'demux'
+
+        Some versions of Clarity do not provide this element when the `DemuxArtifact`
+        is a pool of artifacts that are all from the same submitted sample. In this
+        case, if more than one artifact is labeled, the pool `Artifact`'s demux will
+        be provided, but if only one artifact is labeled, the pool's demux will not
+        be provided.
```
Comment (katherine-epp) [inline s4/clarity/artifact.py:251]: Added this to address the edge case I noticed while writing the comment below.

I'm not sure if it's clear enough that a `DemuxArtifact` that is a pool is a sub-pool in a pool of pools.

**Review by Mark-Swinkels — APPROVED** [review]
